model/flow/event.go

```diff
@@ -99,18 +99,28 @@ type BlockEvents struct {
 
 type EventsList []Event
 
-// EventsListHash calculates a hash of events list,
-// by simply hashing byte representation of events in the lists
-func EventsListHash(el EventsList) (Identifier, error) {
-
-	hasher := hash.NewSHA3_256()
+// EventsListMerkleHash calculates the root hash of events inserted into a
+// merkle trie with the hash of event as the key and encoded event as value
+func EventsListMerkleHash(el EventsList) (Identifier, error) {
+	var root Identifier
+	tree, err := merkle.NewTree(IdentifierLen)
+	if err != nil {
+		return root, err
+	}
 
 	for _, event := range el {
-		_, err := hasher.Write(event.Fingerprint())
+		// event fingerprint is the rlp encoding of the wrapperevent
+		// eventID is the standard sha3 hash of the event fingerprint
+		fingerPrint := event.Fingerprint()
+		eventID := MakeID(fingerPrint)
+		_, err = tree.Put(eventID[:], fingerPrint)
 		if err != nil {
-			return ZeroID, fmt.Errorf("cannot write to hasher: %w", err)
+			return root, err
 		}
 	}
 
-	return HashToID(hasher.SumHash()), nil
+	hash := tree.Hash()
+	copy(root[:], hash)
```

Comment:
As a hot-fix for the spork tomorrow, I was suggesting to work around the edge case by changing the payload tree to always return a 32-byte hash; specifically also in case the tree is empty. I originally implemented it this way but Tarak didn’t like it (see [his comment](https://github.com/onflow/flow-go/pull/1800#discussion_r779167014))

You could pull out my earlier implementation and just re-use the code:
root hash for an empty trie: https://github.com/onflow/flow-go/blob/aa285f72ff6cbb8c601178e909eb9f86bccfe9a9/storage/merkle/tree.go#L428
You might ask yourself, what strange value I was using there for the empty trie (?). It is simply the Blake2B hash of an empty input; in other words, if you instantiate a Blake2B hasher, don’t feed it any values and then compute the Hash, you are getting exactly this constant. 
corresponding test: https://github.com/onflow/flow-go/blob/aa285f72ff6cbb8c601178e909eb9f86bccfe9a9/storage/merkle/node_test.go#L41-L49

Reply:
this was addressed here in [this PR](https://github.com/onflow/flow-go/pull/1833/commits/a038bc26e7990042cb64209ed252dcdf24a8df3e)